### PR TITLE
[graph] add traffic-point animation

### LINKED ISSUE #7934: support trafficPoint animation in PFT graph impl
In Kiali v2 we defer to the built-in PFT animation, which uses animated dashed edges of various speeds. This approach is clean easy to understand, and seems less compute intensive than our prior approach. But it's not as rich, and multiple users have said they prefer our previous "particle" (aka "point") animation, used with cytoscape.

This issue is for offering the old animation as an option in V2 (possibly the default).  I feel like it is doable but I wasn't able to get things working in an initial attempt. I'm not very savvy with thins canvas/context stuff.  Here are some initial findings to possibly help a future try.

The prior animation really doesn't have too much to do with cytoscape itself.  The dots, diamonds, speeds, and density are all created and calculated in kiali code. The animation is rendered independently from the graph, basically rendered on top of the edges, on a separate canvas.  What made this doable was a fairly small cytoscape extension called [cytoscape-canvas](https://github.com/classcraft/cytoscape.js-canvas/blob/master/src/cytoscape-canvas.js), which provided a child-canvas, from which you can get a '2d' context, and render everything directly.

The kiali animation code is found in `TrafficRenderer` and `TrafficPointRenderer`.  In PFT we basically took the code that generates the speeds, and used it to set animation speeds on the PFT edges.  You can see this in `GraphPF`, look for the `Effect` activated by `showTrafficAnimation`.

The question is how to adapt this to PFT. To do so we probably need to again access a 2d canvas we can use to do our rendering. I'm not sure how to do it. If we could get that handle, I'm then not sure whether we would want to render it completely independently, like we do now, or per-edge, possibly adding logic to `StyleEdge`.

PFT has the idea of `Layers`. And you can add custom layers. In an experiment, I created a new ANIMATION_LAYER called "animation", that attempts to be a new layer to which we could write over the graph. In this draft PR you can see a custom layer created here:

https://github.com/kiali/kiali/pull/7933/files#diff-6015c1275b7b3433002b4bd677c35a321c9ff75fd9fa95566b9462551d5b4cfeR368

I can use this layer to successfully write over the graph, you can see in a test here that I'm able to cover the edge in another edge that applies an edge tag of 'foo'.

https://github.com/kiali/kiali/pull/7933/files#diff-084c0ab7a7e65af2ce6d6cddb9bd653c7a340917bfd630cf8833648364527636R169-R192

Also here you can see an attempt to just generate a green rectangle using a 2d context, but it doesn't seem to do anything. I'm stopping, as I've hit a dead end. I'm not sure we need a new Layer, maybe we can somehow just get PFT's top-level container, like cytoscape offers with `cy.container()`, and then just adapt `cytoscape-canvas` (it has open licensing).

### PR DESCRIPTION
**Pull request by jshaughn**
Closes #7934 

Add support to PFT graph for cytoscape-style traffic-point animation.  This PR is based on @ferhoyos approach in https://github.com/kiali/kiali/pull/7943.  This PR now replaces that PR.

The basic approach is to render animated svgelems along with the edge itself.

Operator PR: https://github.com/kiali/kiali-operator/pull/851
- adds setting for which animation style to use

Testing:
@ferhoyos , I think this is ready for review/test.  The default animation is the new TrafficPoint, but you can also revert to the standard PFT animation by setting `kiali_feature_flags.ui_defaults.graph.settings.animation: dash` (there is a operator PR as well, it's already approved),

You can run the Cytoscape graph and the PFT graph side-by-side, and I the results should be fairly similar. I did simplify the error "diamond" so that it is only one element, for me it's OK. Let me know what you think.

To test the parallel edges you'd probably need to install ambient. I have tested it using this latest code.

### CODE REVIEW

**Comment by jshaughn** [comment]
Added bendPoint support, which seems to be working.  But on a separate node, when comparing the pft animation with cytoscape I feel like there are still some minor gaps. The comparative rates don't always seem to be reflected as well.  I will look into it some more.  But we have this!

[bend-anim.webm](https://github.com/user-attachments/assets/fd035f83-264a-40b6-9af6-1b05afc8f499)

File: frontend/src/pages/GraphPF/TrafficAnimation/TrafficRendererPF.tsx [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+      points.unshift(point);
+    }
+
+    return <>{points.map(p => p)}</>;
+  }
+
+  // renewAnimation performs an innocuous "set" to force the edge to re-render and generate new points
+  private renewAnimation(edge: Edge): React.AnimationEventHandler {
+    return _elem => {
+      window.setTimeout(() => {
```
Comment (ferhoyos) [inline frontend/src/pages/GraphPF/TrafficAnimation/TrafficRendererPF.tsx:188]: window is redundant here.
```suggestion
      setTimeout(() => {
```

File: frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    this.animationDuration = animationDuration;
+    this.backgroundColor = backgroundColor;
+    this.borderColor = borderColor;
+    this.percentVisible = percentVisible;
+    this.radius = radius;
+  }
+
+  private getStyle(moveAnimation: string, isInfinite: boolean): string {
+    return kialiStyle({
+      animationDuration: this.animationDuration,
```
Comment (ferhoyos) [inline frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx:139]: The diamond should remain hidden until it starts moving.
```suggestion
      opacity: 0,
      animationDuration: this.animationDuration,
```

File: frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
```
Comment (ferhoyos) [inline frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx:79]: The circle should remain hidden until it starts moving.
```suggestion
      opacity: 0,
      animationDuration: this.animationDuration,
```

File: frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  const startPoint = edge.getStartPoint();
+  const endPoint = edge.getEndPoint();
+  const moveAnimation = {};
+
+  if (edge.getBendpoints().length === 0) {
+    const moveX = endPoint.x - startPoint.x;
+    const moveY = endPoint.y - startPoint.y;
+
+    moveAnimation['0%'] = { translate: '0' };
+    moveAnimation[`${percentVisible}%`] = { translate: `${moveX}px ${moveY}px`, display: isInfinite ? '' : 'none' };
```
Comment (ferhoyos) [inline frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx:25]: The circle and diamond should appear in the graph when they start moving.
```suggestion
    moveAnimation['0%'] = { opacity: 1, translate: '0' };
    moveAnimation[`${percentVisible}%`] = { opacity: 1, translate: `${moveX}px ${moveY}px`, display: isInfinite ? '' : 'none' };
```

File: frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    // a kiali edge can have at most 1 bendpoint, in the middle. see extendedBaseEdge.ts
+    const bendPoint = edge.getBendpoints()[0];
+    const moveBendX = bendPoint.x - startPoint.x;
+    const moveBendY = bendPoint.y - startPoint.y;
+    const moveEndX = endPoint.x - startPoint.x;
+    const moveEndY = endPoint.y - startPoint.y;
+    const bend = Math.round(percentVisible / 2);
+
+    moveAnimation['0%'] = { translate: '0' };
+    moveAnimation[`${bend}%`] = { translate: `${moveBendX}px ${moveBendY}px` };
```
Comment (ferhoyos) [inline frontend/src/pages/GraphPF/TrafficAnimation/TrafficPointRendererPF.tsx:41]: The circle and diamond should appear in the graph when they start moving.
```suggestion
    moveAnimation['0%'] = { opacity: 1, translate: '0' };
    moveAnimation[`${bend}%`] = { opacity: 1, translate: `${moveBendX}px ${moveBendY}px` };
```

**Review by ferhoyos — CHANGES_REQUESTED** [review]
Wow, this is tremendous work, @jshaughn! I’m not entirely sure how to verify all the calculations to ensure everything is working perfectly, hehe.

The point traffic animation is much better than the dash one, and I can understand why some community members still miss the Cytoscape version. I think we’re really close, but there are a few areas that could use some fine-tuning.

One easy suggestion for the code would be to hide the circles and diamonds until they start moving. It looks a bit odd to see all the points waiting in the starting position, like they’re about to race.

Another area to tweak is the uneven spacing of the circles along the edge. Occasionally, the first circle in the cycle starts very close to the last one, which creates the impression of burst traffic. In Cytoscape, the traffic appears to be evenly spaced—at least in the Bookinfo demo—so I’m unsure if this behavior is intentional or not.

Kooha-2024-11-29-14-59-37

Lastly, the traffic animation resets every time the graph is rendered (e.g., when hovering over an edge, the refresh timer ends, etc.). If someone is interacting with the graph, it can feel disruptive for the animation to restart so often. I’m not sure if there’s a workaround for this, but it might be worth looking into.

Kooha-2024-11-29-15-01-25

**Comment by jshaughn** [comment]
> I’m not entirely sure how to verify all the calculations to ensure everything is working perfectly

Right, it took me a while to understand what was happening in the prior animation, and to be honest, I'm using the old animation to sort of validate the new animation.  If they look about the same I think it's good.  They will not look exactly the same because the mechanisms are totally different.  The old approach had the benefit of a separate overlay canvas, and basically could emit traffic-points on timers, adding in some variability etc.  The new approach is more static, we have to generate the svgelems in advance, tie them to the edge rendering, and animate them.  When possible, we infinitely loop the animation, which can reveal patterns.

> The point traffic animation is much better than the dash one, and I can understand why some community members still miss the Cytoscape version. I think we’re really close, but there are a few areas that could use some fine-tuning.

+1, it is much more interesting, gives a better sense of the traffic, and the injection of errors is nice.

> One easy suggestion for the code would be to hide the circles and diamonds until they start moving. It looks a bit odd to see all the points waiting in the starting position, like they’re about to race.

+1. I noticed this as well and wanted to resolve that issue.  I saw you added some suggested changes, I merged them in and will validate it, thanks!

> Another area to tweak is the uneven spacing of the circles along the edge. Occasionally, the first circle in the cycle starts very close to the last one, which creates the impression of burst traffic.

It's not intentional and I'm not sure why it happens or how to avoid it.  I'll recheck the code but the animation-delay is supposed to be even for the generated traffic-points on the edge. It seems like the animation delay is not exactly honored, and is more of a "suggestion".  Remember that here we add all of the svgelems to the edge and it renders as it likes. In cytoscape the trafficpoints are emitted on timers, and independent of the actual graph elements, it's just an overlay.

> Lastly, the traffic animation resets every time the graph is rendered (e.g., when hovering over an edge, the refresh timer ends, etc.). If someone is interacting with the graph, it can feel disruptive for the animation to restart so often. I’m not sure if there’s a workaround for this, but it might be worth looking into.

Again, the problem is that the animation is tied to the edge rendering, and if the edge re-renders the traffic gets re-done.  Remember that on graph refresh the rates can change, so it makes sense to recalculate, but if the rates remain the same it is disruptive.  I'm not sure if there is anything that can be done, because the edge re-renders, but I'll try to make sure we are no re-rendering for no reason.

**Comment by ferhoyos** [comment]
Regarding animation delays, I’ve seen some workarounds to synchronize them (https://stackoverflow.com/questions/4838972/how-to-sync-css-animations-across-multiple-elements). I’ll check if any of them suit our use case.

For the last point, perhaps we could look into using React's memo to prevent the graph component from re-rendering unnecessarily, such as when a refresh timer completes without any changes to the graph.

https://react.dev/reference/react/memo

**Comment by jshaughn** [comment]
@ferhoyos I pushed a change that I think resolves the "reset" issue.  It behaves now pretty much the same as the PFT default dash impl. It detaches the animation rendering from the edge rendering using a memo'd component.  Let me know what you think.

I also pushed a change that I think fixes the "point batching" issue.  I think it had to do with the first point iterating a little too close to the last point launching.

Unfortunately, just as I thought this was done, I discovered a more serious flaw in the logic.  I had come up with what I thought was a nice solution for "injecting" random error points, by forcing re-renders using onAnimationEnd.  This looked great when there was only one point on the edge, like I had in my bookinfo setup using fault injection. But when there are many points on the edge they will all get wiped out when the re-render happens.  I think I'll need to try another approach (this makes me sad because the onAnimationEnd was pretty slick), most likely what it will be will be generating as many traffic points as needed to simulate a "random" number of injected errors, and then iterate infinitely.

**Comment by jshaughn** [comment]
@ferhoyos OK, I updated the error-point-injection approach to be done by generating as many traffic points as needed to simulate a "random" number of injected errors, and then iterating infinitely.  Many points may not actually be visible even though they are still animating.

I don't have anything else pending, so please re-review when you get the chance.

**Review by ferhoyos — CHANGES_REQUESTED** [review]
Traffic animation is in a better shape now. Everything is more fluid and does not re-render every few seconds.

Just two comments:
- If the error rate animation follows the same infinite approach as the success one, is the `IsInfinite` parameter still valid in the code?

- I think we have to fine-tune the `AnimationEdge` component and figure out which fields need to be added to re-render the edge animation. It happens that a change in the error percentage triggers a refresh in the edge, but not in the animation, showing weird things like red diamonds moving in a green edge.

Kooha-2024-12-05-10-17-31

Kooha-2024-12-05-10-18-27

**Comment by jshaughn** [comment]
> If the error rate animation follows the same infinite approach as the success one, is the IsInfinite parameter still valid in the code?

It's not really needed, I left it in there but I can take it out for simplicity.

> I think we have to fine-tune the AnimationEdge component and figure out which fields need to be added to re-render the edge animation. It happens that a change in the error percentage triggers a refresh in the edge, but not in the animation, showing weird things like red diamonds moving in a green edge.

+1, and an excellent catch.  The changes we looks for need to be very fine-grained (not just as an object level), but yes, anything that affects the animation should be considered.  I had previously only considered a position change.  Actually, with the move to 'infinite', I don't think the `animationTime` is still a thing.  I think I can remove that attribute.

File: frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
+import { EdgeAnimationSpeed } from '@patternfly/react-topology';
+import { clamp } from 'utils/MathUtils';
+
+// Some information about the mathematical problem here:
+// We try to project a set of RATE values from an upper-unbounded interval ]0, +Inf[ to a set of TIMER DELAY values in
```
[the remark this answers is not in the document]
Comment (jshaughn) [reply frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts]: You know, I think you have been giving me TOO MUCH credit for this work :-)  I did NOT create AnimationTimerConfig as part of this PR.  This logic is taken from the cytoscape animation (and actually we use it in the PFT dash animation as well, to get the relative animation speeds).  You can see that there is a copy of this file in cytoscape.  I'm sorry if you now think lesser of me :-)

[the remark this answers is not in the document]
Comment (jshaughn) [reply frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts]: I did do a little work to integrate PFT, but the main credit goes originally to our old team members JoseJulio, for the initial impl, and Joel (now on NetObserv), for much of the current logic.

[the remark this answers is not in the document]
Comment (ferhoyos) [reply frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts]: I know that part of the logic is taken from the cytoscape animation, but integrating such complex code with new graph-rendering technology is no easy task, so kudos to everyone!  :raised_hands: (including myself for pointing you toward the path of CSS animation :stuck_out_tongue_winking_eye:)

IIRC, some offset was added to the animated points to indicate traffic deviation? I'm not exactly sure why, but an offset is added in Cytoscape - https://github.com/kiali/kiali/blob/ac2a680831f4bf954c8410194487b3cc679b1544/frontend/src/components/CytoscapeGraph/TrafficAnimation/TrafficRenderer.ts#L380

**Comment by jshaughn** [comment]
@ferhoyos I've made changes to respond to your latest review.  Thanks for sticking with this, hopefully the result will be good.  I also fixed an unrelated issue that I discovered when testing the latest change.  I had added fault injection to change the edge characteristics, and noticed that the VS decorator did not go away when I then deleted the FI.  Now it goes away.

File: frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmazzitelli) [reply frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts]: I remember this... it was for TCP traffic. We were trying to visualize it differently than HTTP so we just offet a little bit (randomly) each dot. Look at TCP traffic in cytoscape (put through lots of traffic to see it) and you will see the offsets.

https://github.com/kiali/kiali/blob/ac2a680831f4bf954c8410194487b3cc679b1544/frontend/src/components/CytoscapeGraph/TrafficAnimation/TrafficRenderer.ts#L148-L153

**Review by ferhoyos — DISMISSED** [review]
Thanks @jshaughn for the quick fix. Now the traffic animation is updated when the error rate changes.

I think the traffic-point animation is mature enough to be merged into master. Well done @jshaughn!

File: frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmazzitelli) [reply frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts]: It is just more visualization clues to the user that they are looking at TCP traffic, not HTTP. I think when you have large amount of data going through, that's when you'll get a good idea what it is doing. Just small amounts of data, it won't be as obvious

[the remark this answers is not in the document]
Comment (jshaughn) [reply frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts]: Oh right, I see now, it's just TCP.  I think is does help differentiate and also to reduce overlap when there are a dense set of traffic points.  Let me see if I can add it in...

[the remark this answers is not in the document]
Comment (jshaughn) [reply frontend/src/pages/GraphPF/TrafficAnimation/AnimationTimerConfig.ts]: We've come this far, may as well try and match the prior animation as much as possible.  Also, when I add this then Mazz can try out the PR as well.

**Comment by jshaughn** [comment]
@jmazzitelli , I added in small offsets for TCP traffic-points, to be like the cytoscape animation.  Can you give this PR a try, just for one more set of eyes outside of myself and @ferhoyos .

**Comment by jmazzitelli** [comment]
Looks OK -- Note that I set RATE to `10000` when installing the bookinfo demo

image

**Comment by jmazzitelli** [comment]
FYI: just looking at top (and using RATE of 100 in bookinfo demo) I am seeing chrome take about 70% CPU utilization for Cytoscape. For PFT, I'm seeing about 95% CPU utilization.

**Comment by jshaughn** [comment]
I feel like we could do a better job of normalizing the number of traffic points.  With rate=100, we're getting really dense dots, and 100 doesn't seem like that much to me.  This is true in Cy as well, both animations are using the same calculations for dots-on-edge.

**Review by jmazzitelli — APPROVED** [review]

**Pull request merged**